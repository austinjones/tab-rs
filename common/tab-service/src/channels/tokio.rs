use crate::{impl_channel_clone, impl_channel_take, Channel};
use tokio::sync::{broadcast, mpsc, oneshot, watch};

impl<T: Send + 'static> Channel for mpsc::Sender<T> {
    type Tx = Self;
    type Rx = mpsc::Receiver<T>;

    fn channel(capacity: usize) -> (Self::Tx, Self::Rx) {
        mpsc::channel(capacity)
    }

    fn default_capacity() -> usize {
        16
    }
}

impl_channel_clone!(mpsc::Sender<T>);
impl_channel_take!(mpsc::Receiver<T>);

impl<T: Send + 'static> Channel for broadcast::Sender<T> {
    type Tx = Self;
    type Rx = broadcast::Receiver<T>;

    fn channel(capacity: usize) -> (Self::Tx, Self::Rx) {
        broadcast::channel(capacity)
    }

    fn default_capacity() -> usize {
        16
    }

    fn clone_rx(rx: &mut Option<Self::Rx>, tx: Option<&Self::Tx>) -> Option<Self::Rx> {
        // tokio channels have a size-limited queue
        // if one receiver stops processing messages,
        // the senders block

        // we take from rx first, getting the bus out of the way
        // then we subscribe using the sender
        // tx should always be here, but just in case.. tx.map( ... )
        rx.take().or_else(|| tx.map(|tx| tx.subscribe()))
    }
}

impl_channel_clone!(broadcast::Sender<T>);

// this is actually overriden in clone_rx
impl_channel_take!(broadcast::Receiver<T>);

impl<T: Send + 'static> Channel for oneshot::Sender<T> {
    type Tx = Self;
    type Rx = oneshot::Receiver<T>;

    fn channel(_capacity: usize) -> (Self::Tx, Self::Rx) {
        oneshot::channel()
    }

    fn default_capacity() -> usize {
        1
    }
}

impl_channel_take!(oneshot::Sender<T>);
impl_channel_take!(oneshot::Receiver<T>);

impl<T> Channel for watch::Sender<T>
where
    T: Default + Clone + Send + Sync + 'static,
{
    type Tx = Self;
    type Rx = watch::Receiver<T>;

    fn channel(_capacity: usize) -> (Self::Tx, Self::Rx) {
        watch::channel(T::default())
    }

    fn default_capacity() -> usize {
        1
    }
}

impl_channel_take!(watch::Sender<T>);
impl_channel_clone!(watch::Receiver<T>);
