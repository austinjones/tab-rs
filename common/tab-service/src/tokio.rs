use crate::Channel;
use tokio::sync::{broadcast, mpsc, oneshot, watch};

pub(crate) fn clone<T: Clone>(link: &mut Option<T>) -> Option<T> {
    link.as_ref().map(|link| link.clone())
}

pub(crate) fn take<T>(link: &mut Option<T>) -> Option<T> {
    link.take()
}

impl<T: 'static> Channel for mpsc::Sender<T> {
    type Tx = Self;
    type Rx = mpsc::Receiver<T>;

    fn channel(capacity: usize) -> (Self::Tx, Self::Rx) {
        mpsc::channel(capacity)
    }

    fn default_capacity() -> usize {
        16
    }

    fn clone_tx(tx: &mut Option<Self::Tx>) -> Option<Self::Tx> {
        tx.as_ref().map(|tx| tx.clone())
    }

    fn clone_rx(rx: &mut Option<Self::Rx>, _: Option<&Self::Tx>) -> Option<Self::Rx> {
        rx.take()
    }
}

impl<T: 'static> Channel for broadcast::Sender<T> {
    type Tx = Self;
    type Rx = broadcast::Receiver<T>;

    fn channel(capacity: usize) -> (Self::Tx, Self::Rx) {
        broadcast::channel(capacity)
    }

    fn default_capacity() -> usize {
        16
    }

    fn clone_tx(tx: &mut Option<Self::Tx>) -> Option<Self::Tx> {
        clone(tx)
    }

    fn clone_rx(rx: &mut Option<Self::Rx>, tx: Option<&Self::Tx>) -> Option<Self::Rx> {
        tx.map(|tx| tx.subscribe())
    }
}

impl<T: 'static> Channel for oneshot::Sender<T> {
    type Tx = Self;
    type Rx = oneshot::Receiver<T>;

    fn channel(capacity: usize) -> (Self::Tx, Self::Rx) {
        oneshot::channel()
    }

    fn default_capacity() -> usize {
        1
    }

    fn clone_tx(tx: &mut Option<Self::Tx>) -> Option<Self::Tx> {
        take(tx)
    }

    fn clone_rx(rx: &mut Option<Self::Rx>, tx: Option<&Self::Tx>) -> Option<Self::Rx> {
        take(rx)
    }
}

impl<T> Channel for watch::Sender<T>
where
    T: Default + Clone + 'static,
{
    type Tx = Self;
    type Rx = watch::Receiver<T>;

    fn channel(capacity: usize) -> (Self::Tx, Self::Rx) {
        watch::channel(T::default())
    }

    fn default_capacity() -> usize {
        1
    }

    fn clone_tx(tx: &mut Option<Self::Tx>) -> Option<Self::Tx> {
        take(tx)
    }

    fn clone_rx(rx: &mut Option<Self::Rx>, tx: Option<&Self::Tx>) -> Option<Self::Rx> {
        take(rx)
    }
}
