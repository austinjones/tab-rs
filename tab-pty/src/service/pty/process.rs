use crate::message::pty::{PtyOptions, PtyRequest, PtyResponse};
use crate::prelude::*;
use std::{
    collections::HashMap,
    process::{Command, ExitStatus},
    sync::Arc,
};
use tab_api::chunk::{InputChunk, OutputChunk};
use tab_pty_process::CommandExt;
use tab_pty_process::{
    AsyncPtyMaster, AsyncPtyMasterReadHalf, AsyncPtyMasterWriteHalf, Child, PtyMaster,
};
use time::Duration;
use tokio::sync::broadcast::RecvError;
use tokio::{
    io::{AsyncReadExt, AsyncWriteExt},
    sync::{
        broadcast::{Receiver, Sender},
        mpsc::error::SendError,
    },
    time,
};

static CHUNK_LEN: usize = 2048;
static OUTPUT_CHANNEL_SIZE: usize = 32;
static STDIN_CHANNEL_SIZE: usize = 32;

#[derive(Debug, Clone)]
pub struct PtyProcess {}

impl PtyProcess {
    fn new() -> PtyProcess {
        PtyProcess {}
    }

    async fn read_output(mut channel: impl AsyncReadExt + Unpin, tx: Sender<PtyResponse>) {
        let mut index = 0usize;
        let mut buffer = vec![0u8; CHUNK_LEN];
        while let Ok(read) = channel.read(buffer.as_mut_slice()).await {
            if read == 0 {
                continue;
            }

            info!("Read {} bytes", read);

            let mut buf = vec![0; read];
            buf.copy_from_slice(&buffer[0..read]);

            let chunk = OutputChunk { index, data: buf };
            let response = PtyResponse::Output(chunk);
            // TODO: deal with error handling
            tx.send(response).expect("Failed to send chunk");
            index += 1;

            // a very short delay allows things to batch up
            // without any buffering, the message rate can get very high
            time::delay_for(Duration::from_millis(5)).await;
        }
    }

    async fn write_input(
        mut stdin: AsyncPtyMasterWriteHalf,
        mut rx: tokio::sync::mpsc::Receiver<PtyRequest>,
    ) {
        while let Some(request) = rx.recv().await {
            match request {
                PtyRequest::Resize(dimensions) => {
                    if let Err(e) = stdin.resize(dimensions).await {
                        error!("failed to resize pty: {:?}", e);
                    }

                    info!("resized to dimensions: {:?}", &dimensions);
                }
                PtyRequest::Input(chunk) => Self::write_stdin(&mut stdin, chunk).await,
                PtyRequest::Shutdown => {
                    info!("terminating pty");
                    stdin.shutdown();
                }
            }
        }

        info!("stdin loop terminated");
    }

    async fn write_stdin(mut stdin: impl AsyncWriteExt + Unpin, mut chunk: InputChunk) {
        info!("writing stdin chunk: {:?}", &chunk);

        // TODO: deal with error handling
        stdin
            .write(chunk.data.as_mut_slice())
            .await
            .expect("Stdin write failed");

        stdin.flush().await.expect("stdin flush failed");
    }
}
