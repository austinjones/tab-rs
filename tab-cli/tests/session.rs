use anyhow::Context;
use insta::assert_debug_snapshot;
use lifeline::assert_completes;
use std::process::Stdio;
use tempfile::tempdir;

use tokio::{io::AsyncReadExt, io::AsyncWriteExt, time};

/// Time to wait for the daemon to launch
const INIT_DELAY_MS: u64 = 1200;

#[tokio::test]
async fn test_session() -> anyhow::Result<()> {
    let dir = tempdir().context("failed to create tempdir")?;
    println!("launching tests in dir: {}", dir.path().to_string_lossy());
    let command = assert_cmd::cargo::cargo_bin("tab");

    let mut run = tokio::process::Command::new(command);
    run.arg("test/session/")
        .env("SHELL", "sh")
        .env("TAB_RUNTIME_DIR", dir.path().to_string_lossy().to_string())
        .env("TAB_RAW_MODE", "false")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .kill_on_drop(true);

    let mut child = run.spawn()?;
    let mut stdin = child.stdin.take().expect("couldn't get child stdin");

    stdin.write_all("echo foo\n".as_bytes()).await?;
    stdin.write_all("exit\n".as_bytes()).await?;
    stdin.flush().await?;

    assert_completes!(
        async move {
            let mut output = child.stdout.take().expect("couldn't get child stdout");
            let mut output_string = "".to_string();
            output
                .read_to_string(&mut output_string)
                .await
                .expect("couldn't read to string");
            assert_debug_snapshot!("result", output_string);
        },
        2000
    );

    Ok(())
}
