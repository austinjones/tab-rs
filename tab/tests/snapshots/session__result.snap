---
source: tab-cli/tests/session.rs
expression: output
---
$ echo foo
foo
$ exit

